### PR TITLE
Add `TRANS_SPRING` interpolation option to tweens

### LINKED ISSUE #4212: Add Spring/Berp Tween transition mode
### Describe the project you are working on

I am working on porting this roguelike tilemap 'game' from Unity to Godot:

- Unity: https://henrysoftware.itch.io/pixel-level (playable)
  - When you drag map off screen it **springs** back.
- Godot: https://henrysoftware.itch.io/godot-pixel-level (playable)
  - When you drag map off screen it **elastics** back.

### Describe the problem or limitation you are having in your project

I miss having access to the Spring tween.
I am not sure where this easing function came from?
I think I first saw it in iTween for Unity. https://github.com/jtothebell/iTween/blob/master/iTween.cs

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It is similar to back or elastic but different enough to notice and seems better suited for ui elements like when OSX and iOS scroll a list past the end and 'spring' back?

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Here is an animation from (https://www.protopie.io/learn/docs/interactions/animation-curves) just to show what it looks like kinda? This one is simpler with no parameters.
[](https://postimg.cc/ctNkMjBD)

You can also see spring tween in action at the PixelLevel link posted above.

Here is a pull request implementing/porting it to Godot.
https://github.com/godotengine/godot/pull/58999

Here is a simple project for testing and comparing new spring tween: https://github.com/rakkarage/TestTween **(4.0 branch!)**

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No.

### Is there a reason why this should be core and not an add-on in the asset library?

Yes, though it is not an official 'Robert Penner' easing function, it is similar in many ways, and many other easing systems include it.

### PR DESCRIPTION
**Pull request by rakkarage**
Thanks for all the help! I hope everything is now in order.
my proposal: https://github.com/godotengine/godot-proposals/issues/4212
my other attempt at a pull request: https://github.com/godotengine/godot/pull/58999

_Production edit: Closes https://github.com/godotengine/godot-proposals/issues/4212_

### CODE REVIEW

File: scene/animation/easing_equations.h [hunk 1]
```diff
@@ -402,4 +402,13 @@ static real_t out_in(real_t t, real_t b, real_t c, real_t d) {
 }
 }; // namespace back
 
+namespace spring {
+static real_t out(real_t t, real_t b, real_t c, real_t d) {
+	t /= d;
+	t = CLAMP(t, 0.0, 1.0);
+	t = (sin(t * Math_PI * (0.2f + 2.5f * t * t * t)) * pow(1.0f - t, 2.2f) + t) * (1.0f + (1.2f * (1.0f - t)));
+	return c * t + b;
+}
```
Comment (TokageItLab) [inline scene/animation/easing_equations.h:411]: Need to implement `in()` /  `in_out()` /  `out_in()` with the same way as Bounce.

**Comment by YuriSizov** [comment]
You used your master branch again... You will likely have issues again when you start making changes after the review or need to rebase.

And with that, please use a more descriptive commit message, e.g. what I changed the title to.

File: scene/animation/tween.cpp [hunk 2]
```diff
@@ -45,6 +45,7 @@ Tween::interpolater Tween::interpolaters[Tween::TRANS_MAX][Tween::EASE_MAX] = {
 	{ &circ::in, &circ::out, &circ::in_out, &circ::out_in },
 	{ &bounce::in, &bounce::out, &bounce::in_out, &bounce::out_in },
 	{ &back::in, &back::out, &back::in_out, &back::out_in },
+	{ &spring::out, &spring::out, &spring::out, &spring::out }, // Spring is the same for each easing.
```
Comment (TokageItLab) [inline scene/animation/tween.cpp:48]: As I said above, I think this comment is probably a wrong. It should be possible to implement in/out/inout/outin the same way as Bounce.

File: scene/animation/easing_equations.h (hunk 1 above)
Comment (rakkarage) [reply scene/animation/easing_equations.h:411]: There are 11 copies of the exact same functions in that file, Perhaps there is a way to separate and reuse...

File: scene/animation/tween.cpp (hunk 2 above)
Comment (KoBeWi) [reply scene/animation/tween.cpp]: Not sure what you mean by "the same as Bounce", but in the [previous PR](https://github.com/godotengine/godot/pull/58999) it was settled that it's ok if spring implements single easing IF it's documented. Not sure if bounce fits here this way.

**Comment by Zireael07** [comment]
github misleading us once again?

**Comment by YuriSizov** [comment]
Yes...

**Pull request merged**